ports/stm32f4/common-hal/pulseio/PWMOut.c

```diff
@@ -0,0 +1,403 @@
+    // mp_printf(&mp_plat_print, "Timer Freq: %d\n", source_freq/(source_freq / PWM_MAX_FREQ));
+    // mp_printf(&mp_plat_print, "Actual Freq: %d\n", (source_freq/(source_freq / PWM_MAX_FREQ))/period);
+    // mp_printf(&mp_plat_print, "Duty: %d\n", duty);
+    // mp_printf(&mp_plat_print, "TIM#:%d CH:%d ALTF:%d\n", self->tim->tim_index, self->tim->channel_index, self->tim->altfn_index);
+
+    //Timer init
+    self->handle.Instance = TIMx;
+    self->handle.Init.Period = period - 1;
+    self->handle.Init.Prescaler = (source_freq / PWM_MAX_FREQ) - 1; // TIM runs at ~6MHz
+    self->handle.Init.ClockDivision = TIM_CLOCKDIVISION_DIV1;
+    self->handle.Init.CounterMode = TIM_COUNTERMODE_UP;
+    self->handle.Init.RepetitionCounter = 0;
+
+    //only run init if this is the first instance of this timer
+    if(first_time_setup) {
+        if(HAL_TIM_PWM_Init(&self->handle) != HAL_OK) {
+            mp_raise_ValueError(translate("Timer Init Error"));
+        }
+    }
+
+    //Channel/PWM init
+    self->chan_handle.OCMode = TIM_OCMODE_PWM1;
+    self->chan_handle.Pulse = (period*duty)/100 - 1;
+    self->chan_handle.OCPolarity = TIM_OCPOLARITY_LOW;
+    self->chan_handle.OCFastMode = TIM_OCFAST_DISABLE;
+    self->chan_handle.OCNPolarity = TIM_OCNPOLARITY_LOW; // needed for TIM1 and TIM8
+    self->chan_handle.OCIdleState = TIM_OCIDLESTATE_SET; // needed for TIM1 and TIM8
+    self->chan_handle.OCNIdleState = TIM_OCNIDLESTATE_SET; // needed for TIM1 and TIM8
+    if(HAL_TIM_PWM_ConfigChannel(&self->handle, &self->chan_handle, self->channel) != HAL_OK) {
+        mp_raise_ValueError(translate("Channel Init Error"));
+    }
+    if(HAL_TIM_PWM_Start(&self->handle, self->channel) != HAL_OK) {
+        mp_raise_ValueError(translate("Error starting PWM"));
+    }
+
+    self->variable_frequency = variable_frequency;
+    self->frequency = frequency;
+    self->duty_cycle = duty;
+
+    return PWMOUT_OK;
+}
+
+bool common_hal_pulseio_pwmout_deinited(pulseio_pwmout_obj_t* self) {
+    return self->tim == mp_const_none;
+}
+
+void common_hal_pulseio_pwmout_deinit(pulseio_pwmout_obj_t* self) {
+    if (common_hal_pulseio_pwmout_deinited(self)) {
+        return;
+    }
+    //var freq shuts down entire timer, others just their channel
+    if(self->variable_frequency) {
+        reserved_tim[self->tim->tim_index-1] = 0x00;
+    } else {
+        reserved_tim[self->tim->tim_index-1] &= ~(1<<self->tim->channel_index);
+        HAL_TIM_PWM_Stop(&self->handle, self->channel);
+    }
+    reset_pin_number(self->tim->pin->port,self->tim->pin->number);
+    self->tim = mp_const_none;
+
+    //if reserved timer has no active channels, we can disable it
+    if (!reserved_tim[self->tim->tim_index-1]) {
+        tim_frequencies[self->tim->tim_index-1] = 0x00;
+        tim_clock_disable(1<<(self->tim->tim_index-1));
+    }
+}
+
+void common_hal_pulseio_pwmout_set_duty_cycle(pulseio_pwmout_obj_t* self, uint16_t duty_cycle) {
+    uint16_t duty = duty_cycle/655;
```

Comment:
where does this 655 magic number come from?

Reply:
I'm struggling with the calculation that generates the duty cycle pulse. The integer math keeps messing up depending on what order the calculation gets done. This was a hack solution (uint16 max 65535/100%) that worked but I'm looking into alternatives.